src/server/worker/datum/util_test.go

```diff
@@ -0,0 +1,67 @@
+package datum
+
+import (
+	"testing"
+
+	"github.com/pachyderm/pachyderm/v2/src/internal/require"
+)
+
+func TestCartesianProduct(t *testing.T) {
+	inputsShards := [][]string{
+		{"a", "b", "c"},
+		{"1", "2"},
+		{"x", "y"},
+	}
+
+	permutations := cartesianProduct(inputsShards, 0)
+	require.Equal(t, 4, len(permutations))
+	expected := [][]string{
+		{"c", "1", "x"},
+		{"c", "1", "y"},
+		{"c", "2", "x"},
+		{"c", "2", "y"},
+	}
+	for _, e := range expected {
```

Comment:
You should rework this test to use `require.NoDiff`.  I think that slicesEqualUnordered and sliceExistsInSlices only exist to get around having to sort expected and expected[], you should just sort those (with cmpopt.SortSlices for []string and [][]string).  The output will be much easier to read and you don't need to maintain these helpers.